lib/src/native/realm_core.dart

```diff
@@ -2645,6 +2712,65 @@ extension LevelExt on Level {
   }
 }
 
+class WriteCompleter with Cancellable implements Completer<void> {
+  final Completer<void> _internalCompleter;
+  final CancellationToken? _cancellationToken;
+  int? _id;
+  final Realm _realm;
+
+  WriteCompleter(this._realm, this._cancellationToken) : _internalCompleter = Completer<void>() {
+    final ct = _cancellationToken;
+    if (ct != null) {
+      if (ct.isCancelled) {
+        _internalCompleter.completeError(CancelledException());
+      } else {
+        ct.attach(this);
+      }
+    }
+  }
+
+  void assignId(int id) {
+    if (_id != null) {
+      throw RealmError('assignId should only be called once');
+    }
+
+    _id = id;
+  }
+
+  /// Whether or not the completer was cancelled.
+  bool get isCancelled => _cancellationToken?.isCancelled ?? false;
+
+  @override
+  bool get isCompleted => _internalCompleter.isCompleted;
+
+  @override
+  Future<void> get future => _internalCompleter.future;
+
+  @override
+  void complete([FutureOr<void>? value]) {
+    if (isCancelled) return;
+    _cancellationToken?.detach(this);
+    _internalCompleter.complete(value);
+  }
+
+  @override
+  void completeError(Object error, [StackTrace? stackTrace]) {
+    if (isCancelled) return;
+    _cancellationToken?.detach(this);
+    _internalCompleter.completeError(error, stackTrace);
+  }
+
+  @override
+  void onCancel(Exception cancelException, [StackTrace? stackTrace]) {
+    if (_id == null || realmCore._cancelAsync(_realm, _id!)) {
```

Comment:
No - if `cancelAsync` returns false, then we should not cancel the future because the cancellation did not succeed. This is typically the case when there's a race between the user calling cancel and Core actually acquiring the write lock. If we complete with an error here, we'd end up in a situation where the dart `beginWrite` is canceled, but the transaction is still open.

Reply:
That is fine though - the completer will resolve normally as if `cancel` was never called. If `_cancelAsync` returns `false`, it means Core didn't register your cancellation attempt and will eventually invoke the callback for the async operation.

Reply:
After a discussion it seems my assumption was incorrect that we want the developer to be able to cancel twice. The thing is if Core returns false on the cancel operation then it will invoke the callback since the operation already completed. By that time it is too late to cancel the operation. So this is a race condition that which was won by Core succeeding the operation and not the user cancelling the op.
So the code is correct here.